File: app/src/main/java/acr/browser/lightning/utils/WebUtils.java
package acr.browser.lightning.utils;

import android.content.Context;
import android.os.Build;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.webkit.CookieManager;
import android.webkit.CookieSyncManager;
import android.webkit.WebStorage;
import android.webkit.WebView;
import android.webkit.WebViewDatabase;

import acr.browser.lightning.database.history.HistoryRepository;
import io.reactivex.Scheduler;

/**
 * Copyright 8/4/2015 Anthony Restaino
 */
public final class WebUtils {

    private WebUtils() {}

    public static void clearCookies(@NonNull Context context) {
        CookieManager c = CookieManager.getInstance();
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            c.removeAllCookies(null);
        } else {
            //noinspection deprecation
            CookieSyncManager.createInstance(context);
            //noinspection deprecation
            c.removeAllCookie();
        }
    }

    public static void clearWebStorage() {
        WebStorage.getInstance().deleteAllData();
    }

    public static void clearHistory(@NonNull Context context,
                                    @NonNull HistoryRepository historyRepository,
                                    @NonNull Scheduler databaseScheduler) {
        historyRepository.deleteHistory()
            .subscribeOn(databaseScheduler)
            .subscribe();
        WebViewDatabase webViewDatabase = WebViewDatabase.getInstance(context);
        webViewDatabase.clearFormData();
        webViewDatabase.clearHttpAuthUsernamePassword();
        Utils.trimCache(context);
    }

    public static void clearCache(@Nullable WebView view) {
        if (view == null) return;
        view.clearCache(true);
    }

}

File: app/src/main/java/acr/browser/lightning/utils/WebUtils.kt
package acr.browser.lightning.utils

import acr.browser.lightning.concurrency.CoroutineDispatchers
import acr.browser.lightning.database.history.HistoryRepository
import android.app.Application
import android.webkit.CookieManager
import android.webkit.WebStorage
import android.webkit.WebView
import android.webkit.WebViewDatabase
import kotlinx.coroutines.withContext
import javax.inject.Inject

class WebUtils @Inject constructor(
    private val application: Application,
    private val coroutineDispatchers: CoroutineDispatchers,
    private val historyRepository: HistoryRepository,
) {
    suspend fun clearCookies() = withContext(coroutineDispatchers.io) {
        CookieManager.getInstance().removeAllCookies(null)
    }

    suspend fun clearWebStorage() = withContext(coroutineDispatchers.io) {
        WebStorage.getInstance().deleteAllData()
    }

    suspend fun clearHistory() = withContext(coroutineDispatchers.io) {
        historyRepository.deleteHistory()
        val webViewDatabase = WebViewDatabase.getInstance(application)
        webViewDatabase.clearFormData()
        webViewDatabase.clearHttpAuthUsernamePassword()
        application.cacheDir.deleteRecursively()
    }

    suspend fun clearCache() = withContext(coroutineDispatchers.main) {
        val webView = WebView(application)
        webView.clearCache(true)
        webView.destroy()
        application.cacheDir.deleteRecursively()
    }
}
